uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js

```diff
@@ -0,0 +1,492 @@
+      $ctrl.balance = {
+        opening: 0,
+        change: {
+          amount: 0,
+          breakdown: []
+        },
+        closing: 0
+      };
+      $ctrl.period = {};
+      $ctrl.pagination = {
+        totalItems: $ctrl.balance.change.breakdown.length,
+        filteredbreakdown: $ctrl.balance.change.breakdown,
+        currentPage: 1,
+        numPerPage: 5,
+        pageChanged: function () {
+          //filter breakdowns
+          var begin = ((this.currentPage - 1) * this.numPerPage),
+            end = begin + this.numPerPage;
+
+          this.filteredbreakdown = $ctrl.balance.change.breakdown.slice(begin, end);
+        }
+      };
+      $ctrl.uiOptions = {
+        showDatePickerFrom: false,
+        showDatePickerTo: false,
+        isChangeExpanded: false,
+        datePickerOptions: {
+          startingDay: 1,
+          showWeeks: false
+        },
+        isAdmin: false,
+        userDateFormat: HR_settings.DATE_FORMAT,
+        showBalance: false,
+        multipleDays: true
+      };
+      $ctrl.error = undefined;
+      $ctrl.loading = {};
+
+      // Create an empty leave request
+      $ctrl.leaveRequest = LeaveRequestInstance.init({
+        contact_id: baseData.contactId //resolved from directive
+      }, false);
+
+      (function () {
+        $ctrl.loading.absenceTypes = true;
+        AbsencePeriod.current().then(function (apInstance) {
+            $ctrl.period = apInstance;
+          })
+          .then(function () {
+            return initAbsenceTypesAndEntitlements().then(function () {
+              $ctrl.loading.absenceTypes = false;
+            });
+          })
+          .then(function () {
+            return initDayTypesAndStatus()
+          });
+      })();
+
+      /**
+       * Initializes values for absence types and entitlements when the model is loaded
+       *
+       * @returns {Promise}
+       */
+      function initAbsenceTypesAndEntitlements() {
+        // Fetch all the absence types, except for the sickness ones
+        return AbsenceType.all({
+            is_sickness: false
+          })
+          .then(function (absenceTypes) {
+            var absenceTypesIds = absenceTypes.map(function (item) {
+              return item.id;
+            })
+
+            // And then for each of them get the remaining balance from the
+            // entitlements linked to them
+            Entitlement.all({
+                contact_id: $ctrl.leaveRequest.contact_id,
+                period_id: $ctrl.period.id,
+                type_id: { in: absenceTypesIds
+                }
+              }, true) // `true` because we want to use the 'future' balance for calculation
+              .then(function (entitlements) {
+                // create a list of absence types with a `balance` property
+                $ctrl.absenceTypes = entitlements.map(function (entitlementItem) {
+                  var absenceType = absenceTypes.find(function (absenceTypeItem) {
```

Comment:
[`Array.prototype.find()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) is not supported by IE